services/api-server/src/simcore_service_api_server/services/log_streaming.py

```diff
@@ -0,0 +1,135 @@
+import asyncio
+from asyncio import Queue
+from typing import AsyncIterable, Awaitable, Callable, Final
+
+from models_library.rabbitmq_messages import LoggerRabbitMessage
+from models_library.users import UserID
+from pydantic import PositiveInt
+from servicelib.rabbitmq import RabbitMQClient
+
+from ..models.schemas.jobs import JobID, JobLog
+from .director_v2 import DirectorV2Api
+
+_NEW_LINE: Final[str] = "\n"
+_SLEEP_SECONDS_BEFORE_CHECK_JOB_STATUS: Final[PositiveInt] = 10
+
+
+class LogDistributionBaseException(Exception):
+    pass
+
+
+class LogStreamerNotRegistered(LogDistributionBaseException):
+    pass
+
+
+class LogStreamerRegistionConflict(LogDistributionBaseException):
+    pass
+
+
+class LogDistributor:
+    def __init__(self, rabbitmq_client: RabbitMQClient):
+        self._rabbit_client = rabbitmq_client
+        self._log_streamers: dict[JobID, Callable[[JobLog], Awaitable[None]]] = {}
+        self._queue_name: str
+
+    async def setup(self):
+        self._queue_name = await self._rabbit_client.subscribe(
+            LoggerRabbitMessage.get_channel_name(),
+            self._distribute_logs,
+            exclusive_queue=True,
+            topics=[],
+        )
+
+    async def teardown(self):
+        await self._rabbit_client.unsubscribe(self._queue_name)
+
+    async def __aenter__(self):
+        await self.setup()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.teardown()
+
+    async def _distribute_logs(self, data: bytes):
+        got = LoggerRabbitMessage.parse_raw(data)
+        item = JobLog(
+            job_id=got.project_id,
+            node_id=got.node_id,
+            log_level=got.log_level,
+            messages=got.messages,
+        )
+        callback = self._log_streamers.get(item.job_id)
+        if callback is None:
+            raise LogStreamerNotRegistered(
```

Comment:
Question:
Do I get that right that if you raise here the message is sent back to rabbit which will eventually present it again to the very same api-server. is this what you want? just wondering whether that make sense.

Reply:
No, the exceptions which can be raised by the `LogDistributor` or `LogStreamer` are handled [here](https://github.com/bisgaard-itis/osparc-simcore/blob/dba4b448ea76c632f8de2cf38c4fec1579e6f751/services/api-server/src/simcore_service_api_server/api/errors/log_handling_error.py#L13-L14) and returned to the client which made the request as an appropriate HTTPException. So nothing goes back into the rabbit.